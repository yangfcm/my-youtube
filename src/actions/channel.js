import { defAxios as axios } from "../settings";
import {
  FETCH_CHANNEL,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_CHANNEL_INTRO,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
  CATCH_ERROR,
} from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";
