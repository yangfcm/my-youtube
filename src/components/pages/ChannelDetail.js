import React from "react";
import { Switch, Route, Redirect } from "react-router-dom";
import Banner from "../layout/Banner";
import Menu from "../layout/Menu";

import ChannelIntro from "./ChannelIntro";
import ChannelVideos from "./ChannelVideos";
import ChannelPlaylist from "./ChannelPlaylist";
import ChannelBanner from "../layout/ChannelBanner";

class ChannelDetail extends React.Component {
  render() {
    const { path, url } = this.props.match;
    const channelMenuItems = [
      { link: `${url}/videos`, exact: false, name: "Videos", icon: "video" },
      { link: `${url}/playlist`, exact: false, name: "Playlist", icon: "list" },
      {
        link: `${url}/intro`,
        exact: false,
        name: "Channel Intro",
        icon: "info-circle",
      },
    ];
    return (
      <div>
        <ChannelBanner channelId={this.props.match.params.id} />
        <Menu menuItems={channelMenuItems} />
        <Switch>
          <Route path={`${path}/videos`}>
            <ChannelVideos channelId={this.props.match.params.id} />
          </Route>
          <Route path={`${path}/playlist`}>
            <ChannelPlaylist channelId={this.props.match.params.id} />
          </Route>
          <Route path={`${path}/intro`}>
            <ChannelIntro channelId={this.props.match.params.id} />
          </Route>
          <Route path={`${path}`} exact>
            <Redirect to={`${url}/videos`} />
          </Route>
          <Route path={`${path}/*`}>
            <Redirect to="/not-found" />
          </Route>
        </Switch>
      </div>
    );
  }
}

export default ChannelDetail;
