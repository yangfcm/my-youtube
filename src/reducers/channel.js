import {
  FETCH_CHANNEL,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_CHANNEL_INTRO,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
} from "../actions/types";

export const channelReducer = (state = {}, action) => {
  switch (action.type) {
    case FETCH_CHANNEL:
      return {
        ...state,
        channels: action.payload,
      };
    case FETCH_CHANNEL_INTRO:
      return {
        ...state,
        channelIntro: action.payload,
      };
    case FETCH_CHANNEL_SUBSCRIPTION:
      return {
        ...state,
        isSubscribed: action.payload.isSubscribed,
      };
    case SUBSCRIBE_CHANNEL:
      return {
        ...state,
        subscribedChannel: action.payload,
      };
    case UNSUBSCRIBE_CHANNEL:
      return {
        ...state,
        subscribedChannel: null,
      };
    default:
      return state;
  }
};
