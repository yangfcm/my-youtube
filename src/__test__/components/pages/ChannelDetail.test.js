import React from "react";
import { shallow } from "enzyme";
import configMockStore from "redux-mock-store";
import ChannelDetail from "components/pages/ChannelDetail";
import { channelIntroData } from "__test__/fixtures/channel";

describe("Test ChannelDetail component", () => {
  let wrapper, store, match;
  const channelId = "channel_id";
  beforeEach(() => {
    const mockStore = configMockStore([]);
    store = mockStore({
      channel: {
        channelIntro: channelIntroData,
      },
      search: {
        searchResults: {},
      },
    });
    store.dispatch = jest.fn();
  });

  afterEach(() => {
    wrapper.unmount();
  });

  describe("Test component at route /channel/:id", () => {
    beforeEach(() => {
      match = {
        params: { id: channelId },
        path: "/channel/:id",
        url: `/channel/${channelId}`,
        isExact: true,
      };
      wrapper = shallow(<ChannelDetail match={match} />);
    });

    it("should render component", () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
