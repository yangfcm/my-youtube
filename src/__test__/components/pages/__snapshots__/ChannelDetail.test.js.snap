// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChannelDetail component Test component at route /channel/:id should render component 1`] = `
<div>
  <Connect(withRouter(ChannelBanner))
    channelId="channel_id"
  />
  <Menu
    menuItems={
      Array [
        Object {
          "exact": false,
          "icon": "video",
          "link": "/channel/channel_id/videos",
          "name": "Videos",
        },
        Object {
          "exact": false,
          "icon": "list",
          "link": "/channel/channel_id/playlist",
          "name": "Playlist",
        },
        Object {
          "exact": false,
          "icon": "info-circle",
          "link": "/channel/channel_id/intro",
          "name": "Channel Intro",
        },
      ]
    }
  />
  <Switch>
    <Route
      path="/channel/:id/videos"
    >
      <Connect(ChannelVideos)
        channelId="channel_id"
      />
    </Route>
    <Route
      path="/channel/:id/playlist"
    >
      <Connect(ChannelPlaylist)
        channelId="channel_id"
      />
    </Route>
    <Route
      path="/channel/:id/intro"
    >
      <Connect(withRouter(ChannelIntro))
        channelId="channel_id"
      />
    </Route>
    <Route
      exact={true}
      path="/channel/:id"
    >
      <Redirect
        to="/channel/channel_id/videos"
      />
    </Route>
    <Route
      path="/channel/:id/*"
    >
      <Redirect
        to="/not-found"
      />
    </Route>
  </Switch>
</div>
`;
